lib/libxdp/xsk.c

```diff
@@ -696,6 +780,34 @@ static int __xsk_setup_xdp_prog(struct xsk_socket *xsk, int *xsks_map_fd)
 	if (IS_ERR(ctx->xdp_prog))
 		return PTR_ERR(ctx->xdp_prog);
 
+	if (ctx->xdp_prog) {
+		int refcnt;
+
+		ctx->refcnt_map_fd = xsk_lookup_refcnt_map(xdp_program__fd(ctx->xdp_prog));
+		if (ctx->refcnt_map_fd == -ENOENT) {
+			pr_warn("AF_XDP refcount map was not found. If you want to use built-in refcount, please recreate your sockets with XDP program that contains int variable in .data section\n");
+			ctx->refcnt_map_fd = 0;
+			goto map_lookup;
+		}
+
+		if (ctx->refcnt_map_fd < 0) {
+			err = ctx->refcnt_map_fd;
+			pr_warn("Error getting refcount map\n");
+			ctx->refcnt_map_fd = 0;
+			goto err_prog_load;
+		}
+
+		refcnt = xsk_incr_prog_refcnt(ctx->refcnt_map_fd);
+		if (refcnt < 2) {
+			pr_warn("Could not update current refcount: %d, falling back on creating a new program\n",
+				refcnt);
```

Comment:
So the fallback is obviously the right thing to do if we end up racing with another program that is tearing down (i.e., if refcnt is 0). But if we get an error on map update, should we really just continue?